File: 프로그래머스/1/92334. 신고 결과 받기/신고 결과 받기.cs
using System;
using System.Collections.Generic;

public class Solution {
    public int[] solution(string[] id_list, string[] report, int k) {
        int[] answer = new int[id_list.Length];
        
        var index = new Dictionary<string, int>();
        for(int i = 0; i < id_list.Length; i++)
        {
            index[id_list[i]] = i;
        }
        
        var reportsBy = new Dictionary<string, HashSet<string>>();
        var reportedBy = new Dictionary<string, HashSet<string>>();

        
        foreach(var id in id_list)
        {
            reportsBy[id] = new HashSet<string>();
            reportedBy[id] = new HashSet<string>();
        }
        
        foreach(var s in report)
        {
            var parts = s.Split(' ');
            string reporter = parts[0];
            string target = parts[1];
            
            if(reportsBy[reporter].Add(target))
            {
                reportedBy[target].Add(reporter);
            }
        }
        
        var suspended = new HashSet<string>();
        foreach(var id in id_list)
        {
            if(reportedBy[id].Count >= k) suspended.Add(id);
        }
        
        foreach(var reporter in id_list)
        {
            int mailCount = 0;
            foreach(var target in reportsBy[reporter])
            {
                if(suspended.Contains(target))
                    mailCount++;
            }
            answer[index[reporter]] = mailCount;
        }
        return answer;
    }
}

File: 프로그래머스/1/92334. 신고 결과 받기/신고 결과 받기.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

vector<int> solution(vector<string> id_list, vector<string> report, int k) {
    int n = id_list.size();
    vector<int> answer(n, 0);
    
    set<string> s(report.begin(), report.end());
    map<string, int> cnt;
    map<string, int> idx;
    
    for(int i = 0; i < n; i++)
    {
        idx[id_list[i]] = i;
    }
    
    for(string str : s)
    {
        stringstream ss(str);
        string from, to;
        
        ss >> from >> to;
        
        cnt[to]++;
    }
    
    for(string str : s)
    {
        stringstream ss(str);
        string from, to;
        
        ss >> from >> to;
        
        if(cnt[to] >= k)
        {
            answer[idx[from]]++;
        }
    }
    
    return answer;
}